Tests/CommandsTests/BuildToolTests.swift

```diff
@@ -111,11 +111,89 @@ final class BuildToolTests: XCTestCase {
         }
     }
 
+    func testNonReachableProductsAndTargets() throws {
+        // Dependency contains a dependent product
+
+        fixture(name: "Miscellaneous/UnreachableTargets") { path in
+            do {
+                let aPath = path.appending(component: "A")
+                let output = try execute(["--product", "bexec"], packagePath: aPath)
+                XCTAssert(output.contains("Compile") && output.contains("BTarget2"))
+                XCTAssert(output.contains("Linking") && output.contains("bexec"))
+                XCTAssert(!output.contains("aexec"))
+                XCTAssert(!output.contains("ATarget"))
+                XCTAssert(!output.contains("BLibrary"))
+                XCTAssert(!output.contains("BTarget1"))
+                XCTAssert(!output.contains("cexec"))
+                XCTAssert(!output.contains("CTarget"))
+            } catch SwiftPMProductError.executionFailure(_, _, let stderr) {
+                XCTFail(stderr)
+            }
+        }
+
+        fixture(name: "Miscellaneous/UnreachableTargets") { path in
+            do {
+                let aPath = path.appending(component: "A")
+                let output = try execute(["--target", "BTarget2"], packagePath: aPath)
+                print("output: \(output)")
+                XCTAssert(output.contains("Compile") && output.contains("BTarget2"))
+                XCTAssert(!output.contains("Linking"))
+                XCTAssert(!output.contains("aexec"))
+                XCTAssert(!output.contains("ATarget"))
+                XCTAssert(!output.contains("BLibrary"))
+                XCTAssert(!output.contains("bexec"))
+                XCTAssert(!output.contains("BTarget1"))
+                XCTAssert(!output.contains("cexec"))
+                XCTAssert(!output.contains("CTarget"))
+            } catch SwiftPMProductError.executionFailure(_, _, let stderr) {
+                XCTFail(stderr)
+            }
+        }
+
+        // Dependency does not contain a dependent product
+
+        fixture(name: "Miscellaneous/UnreachableTargets") { path in
+            do {
+                let aPath = path.appending(component: "A")
+                let output = try execute(["--product", "cexec"], packagePath: aPath)
+                XCTAssert(output.contains("Compile") && output.contains("CTarget"))
+                XCTAssert(output.contains("Linking") && output.contains("cexec"))
+                XCTAssert(!output.contains("aexec"))
+                XCTAssert(!output.contains("ATarget"))
+                XCTAssert(!output.contains("BLibrary"))
+                XCTAssert(!output.contains("bexec"))
+                XCTAssert(!output.contains("BTarget1"))
+                XCTAssert(!output.contains("BTarget2"))
```

Comment:
I'll land some APIs for string pattern matching. That should help a lot in this type of asserts.